appsrc/dcfile/dctopnm.man
.TH DCTOPNM 1 "6 March 2014" "DICOM PS3" "DICOM PS3 - DICOM image to PNM file"
.SH NAME
dctopnm \- ACR/NEMA DICOM PS3 ... DICOM PS3 - DICOM image to PNM file
.SH SYNOPSIS
.HP 10
.B dctopnm
.so man1/gen.so
[
.B \-v|verbose
]
[
.B \-quiet|silent
]
.so man1/optin.so
.so man1/binout.so
.SH DESCRIPTION
.LP
.B dctopnm
reads the named dicom or acr-nema input file and copies the raw image
pixel data to a PNM file after adding a PGM or PPM header.
.LP
The input file must be unencapsulated, single channel (grayscale or palette color; the photometric interpretation is not checked for single channel), or three channel interleaved RGB, and in the case of > 8 bit data,
in the same byte order as desired in the output, since the raw bytes
are copied from input to output without any conversion. Use dccp(1)
first to change the encoding if necessary.
.LP
Only the necessary number of bytes
to encode the pixel data will be written (based on rows, columns, number of
frames, samples per pixel and bits allocated); this means that for an odd number
of bytes the extra DICOM padding byte will not be copied to the raw output. This
prevents netpbm utilities freaking out with an error "reading magic number" if the padding byte is
present. If
any of the necessary attributes to calculate
the correct pixel data length are missing, all encoded bytes will be written.
.LP
.SH OPTIONS
The PNM output goes where it is specified or standard out.
The verbose output goes to standard error.
.LP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-quiet|silent
.RS
Suppress the normal description of image parameters.
.RE
.TP
.B \-verbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading, once read, during replacement, and before writing.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% dcfile big.dc3
.RE
Meta: UID		1.2.840.10008.1.2.1
.RE
Meta: Description	"Explicit VR Little Endian"
.RE
Meta: ByteOrder	Little
.RE
Meta: VR		Explicit
.RE
Meta: Encapsulated	No
.RE
Data: UID		1.2.840.10008.1.2.2
.RE
Data: Description	"Explicit VR Big Endian"
.RE
Data: ByteOrder	Big
.RE
Data: VR		Explicit
.RE
Data: Encapsulated	No
.RE
\ 
.RE
% dctopnm big.dc3 big.pgm -endian big
.RE
 Rows = 512
.RE
 Columns = 512
.RE
 NumberOfFrames = 0
.RE
 PhotometricInterpretation = MONOCHROME2
.RE
 SamplesPerPixel = 1
.RE
 BitsAllocated = 16
.RE
 BitsStored = 16
.RE
 HighBit = 15
.RE
 PixelRepresentation = 1
.RE
 PlanarConfiguration = 0x0
.RE
 Rows*Columns*SamplesPerPixel*BitsAllocated/8 = 0x80000
.RE
Writing ...
.RE
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dccp(1) ,
.BR pnmtodc(1) ,
.BR rawtodc(1) ,
.BR dctoraw(1) ,
.BR dctopgm8(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (c) 1993-2020, David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS
It would be desirable to be able to expand palette color images to produce RGB output, and to handle color-by-plane rather than interleaved RGB, but for now, these are not supported.


CHANGES
210306:	libsrc/standard/condn.tpl,iodcomp,module,strval/vl.tpl: Add Dermoscopic Photography Image (Sup 221)

210306:	libsrc/standard/binval.tpl,iodcomp,module,strval/vl.tpl: Add Visible Light Photography Attributes corresponding to EXIF 2.31 and TIFF/EP (CP 1736)

210206:	libsrc/standard/condn.tpl: Include SCT codes when determining Laterality Required, using 2021a as refernce, including some updated codes (000555)

210121:	libsrc/standard/elmdict/dicom3.tpl: Add new data elements for XA Protocol (Sup 212)

210119:	libsrc/standard/elmdict/dicom3.tpl: Add new data elements (CPACK 109)

210104:	libsrc/standard/elmdict/other.tpl: Update PixelMed private WSI bulk annotation data elements with Annotation Group UID (sup222_08_WSIAnnotations)

210102:	libsrc/standard/elmdict/dicom3.tpl,module/rt.tpl: Source Pixel Plane Characteristics for Contours (CP 2006); Extend RT Structure Set ROI Color (CP 1395)

201208:	appsrc/dcfile/dciodvfy.cc: Check sanity of tile dimensions and number of frames in whole slide images (000553)

201208: libsrc/standard/elmdict/other.tpl: Add more observed 3DHISTECH private data elements

201208: libsrc/standard/condn.tpl: Do not report zero length ViewModifierCodeSequence error for non Mammo DX when ViewModifierCodeSequence not present (regression caused by 000549) (000552)

201208: libsrc/standard/condn.tpl: Do not complain about presence of ReferencedSeriesSequence or StudiesContainingOtherReferencedInstancesSequence (Common Instance Reference Module) if SegmentSurfaceSourceInstanceSequence is present (000551)

201208: libsrc/src/dctool/condn.cc,standard/condn.tpl,support/condn.awk: Do not complain about absence of ReferencedSeriesSequence and StudiesContainingOtherReferencedInstancesSequence (Common Instance Reference Module) when valid empty (no item) ReferencedImageSequence (000550)

201205: libsrc/standard/condn.tpl,module/dx.tpl: Do not report error on zero length ViewModifierCodeSequence if Mammo, since Type 2 specializes DX Positioning Module Type 3 (regression caused by change on 2013/06/26) (000549)

201117:	libsrc/standard/elmdict/dicom3.tpl: Add OV, SV and UV Selector Attributes (CP 2042)

201117:	libsrc/standard/elmdict/dicom3.tpl: SR Table data elements (CP 2041)

201115:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Dermoscopy data elements (Sup 221)

200920:	standard/module/base.tpl,module/base.tpl: Make Frame Anatomy in Parametric Map Optional (CP 2004)

200919:	libsrc/standard/condn.tpl,elmdict/dicom3.tpl,module/base.tpl,us.tpl: Add ultrasound Probe ID in Transducer Identification Sequence (CP 1951); depended on adding Device Identification Sequence Macro (from Sup 147)

200915:	libsrc/standard/elmdict/other.tpl: Update PixelMed private WSI bulk annotation data elements (supnnn_05_WSIAnnotations)

200829:	libsrc/standard/condn.tpl,iodcomp,module/base.tpl: Check CommonInstanceReference Module presence conditional on references being present (000547) and attributes only present when references are present (000548)

200822:	libsrc/standard/elmdict/dicom3.tpl: Render MakeNote as string even though OB

200806:	standard/condn.tpl,module/base.tpl,mr.tpl,vl.tpl: Do not complain about YBR_RCT for .90 Transfer Syntax (000546)

200726:	libsrc/standard/strval/vl.tpl: Correct NUMERIC value type in Acquisition Context (000545)

200724:	libsrc/standard/module/mr.tpl: Fix incorrect condition for Real World Value Mapping Macro in Shared Functional Group Sequence for Legacy Converted MR (000544)

200719:	libsrc/standard/condn.tpl,elmdict/dicom3.tpl,diconde.tpl: Correct VRs, VMs and keywords to match 2020c (Peros)

200716:	libsrc/standard/elmdict/siemens.tpl: Correct duplicate and incorrect private data elements (Vrinda, Conformance Statement)

200716:	libsrc/standard/elmdict/gems.tpl: Correct duplicate and incorrect private data elements (Vrinda, Conformance Statement)

200716:	libsrc/standard/elmdict/agfa.tpl: Correct duplicate and incorrect private data elements (Vrinda, Conformance Statement)

200716:	libsrc/standard/condn.tpl: Correct KVP CT Image Module condition when multi-energy (000543)

200716:	libsrc/standard/module/ct.tpl: Use ReferencedXRaySourceIndex not ReferencedPathIndex for multi-energy CT Exposure Sequence (000542)

200716:	libsrc/standard/elmdict/dicom3.tpl: Add data elements and SOP Class UIDs for latest release (includng catching up on 2nd Gen RT UIDs) (2020c)

200622:	libsrc/include,src/dctool/attrtypo.h,.cc: Allow insertion of strings in OB using addValues(), e.g., for dccp (000541)

200620:	libsrc/src/dctool/dcopto.cc: Allow insertion of empty sequence during dccp (000540)

200610:	libsrc/standard/elmdict/philips.tpl,siemens.tpl: Update observed private data elements, including OB rendered as string for some

200610:	libsrc/include/dctool/ie.h,src/dctool/ie.cc,support/module.awk: Fix dcentvfy SeriesInstanceUID broken by addition of ReferencedInstancesAndAccessMacro for Patient Photo by overwriting IE for attribute set during module parsing when depth in the hierarchy is lower (000539)

200608:	libsrc/standard/condn.tpl,elmdict/dicom3.tpl,module,strval/base.tpl: Update Encapsulated Document Module (CP 1811 and Sup 208, including new data elements)

200605:	libsrc/standard/elmdict/other.tpl: Add Merge Unity PACS (DR Systems) data elements as described in DCS v11.2 and encountered

200605:	libsrc/standard/elmdict/other.tpl: Add Philips/Stentor iSite data elements as encountered

200604:	libsrc/standard/elmdict/other.tpl: Add Samsung CR private data elements as encountered, as UT

200604:	libsrc/standard/elmdict/gems.tpl: More private GEMS_PARM_01 group 0043 GE MR data elements from DV26, PX26.0, SV25.1, PX26.1 DCS

200526:	libsrc/standard/elmdict/elscint.tpl: Correct VR of (5001,0070) from SQ to US and other related elements based on observed GSPS - may be conflict with ELSCINT1 creator [fragile]

200520:	libsrc/standard/elmdict/toshiba.tpl: Add more private data elements from CT Conformance Statements


appsrc/dcfile/dctopnm.cc
static const char *CopyrightIdentifier(void) { return "@(#)dctopnm.cc Copyright (c) 1993-2020, David A. Clunie DBA PixelMed Publishing. All rights reserved."; }
#include "attrmxls.h"
#include "attrothr.h"
#include "attrval.h"
#include "mesgtext.h"
#include "bnopt.h"
#include "ioopt.h"
#include "dcopt.h"
#include "elmconst.h"
#include "transynd.h"
#include "transyn.h"

int
main(int argc, char *argv[])
{
	GetNamedOptions 			options(argc,argv);
	DicomInputOptions 			dicom_input_options(options);
	BinaryOutputOptionsWithByteOrder	output_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool quiet=options.get("quiet") || options.get("silent");

	dicom_input_options.done();
	output_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);
	OutputOpenerFromOptions output_opener(
		options,output_options.filename,cout);

	cerr << dicom_input_options.errors();
	cerr << output_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();
	cerr << output_opener.errors();

	if (!dicom_input_options.good()
	 || !output_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !output_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< output_options.usage()
			<< " [-v|-verbose]"
			<< " [-quiet|-silent]"
			<< " [" << MMsgDC(InputFile)
				<< "[" << MMsgDC(OutputFile) << "]]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);
	ostream out(output_opener);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream log(cerr);
	if (verbose) log << "******** While reading ... ********" << endl; 
	list.read(din,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	const char *errors=list.errors();
	if (errors) log << errors << flush;
	if (!list.good()) {
		log << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}

	Uint16 vRows = 0;
	Attribute *aRows = list[TagFromName(Rows)];
	if (!aRows)
		log << WMsgDC(MissingAttribute)
		    << " - \"Rows\""
		    << endl;
	else
		vRows=AttributeValue(aRows);

	Uint16 vColumns = 0;
	Attribute *aColumns = list[TagFromName(Columns)];
	if (!aColumns)
		log << WMsgDC(MissingAttribute)
		    << " - \"Columns\""
		    << endl;
	else
		vColumns=AttributeValue(aColumns);

	Uint16 vNumberOfFrames = 0;
	Attribute *aNumberOfFrames = list[TagFromName(NumberOfFrames)];
	if (aNumberOfFrames)	// optional
		vNumberOfFrames=AttributeValue(aNumberOfFrames);

	if (vNumberOfFrames == 0) {
		vNumberOfFrames = 1;	// need to treat it as 1 if missing for later byteCount calculation
	}

	char *vPhotometricInterpretation = 0;
	Attribute *aPhotometricInterpretation = list[TagFromName(PhotometricInterpretation)];
	if (!aPhotometricInterpretation)
		log << WMsgDC(MissingAttribute)
		    << " - \"PhotometricInterpretation\""
		    << endl;
	else
		vPhotometricInterpretation=AttributeValue(aPhotometricInterpretation);

	Uint16 vSamplesPerPixel = 0;
	Attribute *aSamplesPerPixel = list[TagFromName(SamplesPerPixel)];
	if (!aSamplesPerPixel)
		log << WMsgDC(MissingAttribute)
		    << " - \"SamplesPerPixel\""
		    << endl;
	else
		vSamplesPerPixel=AttributeValue(aSamplesPerPixel);

	Uint16 vBitsAllocated = 0;
	Attribute *aBitsAllocated = list[TagFromName(BitsAllocated)];
	if (!aBitsAllocated)
		log << WMsgDC(MissingAttribute)
		    << " - \"BitsAllocated\""
		    << endl;
	else
		vBitsAllocated=AttributeValue(aBitsAllocated);

	Uint16 vBitsStored = 0;
	Attribute *aBitsStored = list[TagFromName(BitsStored)];
	if (!aBitsStored)
		log << WMsgDC(MissingAttribute)
		    << " - \"BitsStored\""
		    << endl;
	else
		vBitsStored=AttributeValue(aBitsStored);

	Uint16 vHighBit = 0;
	Attribute *aHighBit = list[TagFromName(HighBit)];
	if (!aHighBit)
		log << WMsgDC(MissingAttribute)
		    << " - \"HighBit\""
		    << endl;
	else
		vHighBit=AttributeValue(aHighBit);

	Uint16 vPixelRepresentation = 0xffff;
	Attribute *aPixelRepresentation = list[TagFromName(PixelRepresentation)];
	if (!aPixelRepresentation)
		log << WMsgDC(MissingAttribute)
		    << " - \"PixelRepresentation\""
		    << endl;
	else
		vPixelRepresentation=AttributeValue(aPixelRepresentation);

	Uint16 vPlanarConfiguration = 0xffff;
	Attribute *aPlanarConfiguration = list[TagFromName(PlanarConfiguration)];
	if (vSamplesPerPixel > 1 && !aPlanarConfiguration)
		log << WMsgDC(MissingAttribute)
		    << " - \"PlanarConfiguration\""
		    << endl;
	else
		vPlanarConfiguration=AttributeValue(aPlanarConfiguration);

	Uint32 byteCount = ((Uint32)vRows) * vColumns * vNumberOfFrames * vSamplesPerPixel * ((vBitsAllocated-1)/8 + 1);

	if (!quiet) {
		log << "\tRows = " << dec << vRows << endl;
		log << "\tColumns = " << dec << vColumns << endl;
		log << "\tNumberOfFrames = " << dec << vNumberOfFrames << endl;
		log << "\tPhotometricInterpretation = "
		    << (vPhotometricInterpretation ? vPhotometricInterpretation : "")
		    << endl;
		log << "\tSamplesPerPixel = " << dec << vSamplesPerPixel << endl;
		log << "\tBitsAllocated = " << dec << vBitsAllocated << endl;
		log << "\tBitsStored = " << dec << vBitsStored << endl;
		log << "\tHighBit = " << dec << vHighBit << endl;
		log << "\tPixelRepresentation = " << dec << vPixelRepresentation << endl;
		log << "\tPlanarConfiguration = " << hex << vPlanarConfiguration << dec << endl;
		log << "\tRows*Columns*vNumberOfFrames*SamplesPerPixel*((vBitsAllocated-1)/8+1) = "
		    << hex << byteCount
		    << dec << " (" << byteCount << " dec)" << endl;
	}

	if (vNumberOfFrames > 1) {
		log << "Error - Multiple frames not supported" << endl;
		success=false;
	}

	if (din.getTransferSyntaxToReadDataSet()->isEncapsulated()) {
		log << "Error - Encapsulated (compressed) input data not supported" << endl;
		success=false;
	}

	if (vBitsAllocated > 8) {
		if (output_options.byteorder == ByteEndian || output_options.byteorder == NoEndian) {
			log << "Error - Need to specify output byte order for > 8 bit data" << endl;
			success=false;
		}
		else if (din.getEndian() != output_options.byteorder) {
			log << "Error - Need to specify same byte order as input for > 8 bit data" << endl;
			success=false;
		}
	}

	const char *pnmstring=0;

	if (vSamplesPerPixel == 1) {
		pnmstring="P5";
	}
	else if (vSamplesPerPixel == 3
		&& vPhotometricInterpretation && strcmp(vPhotometricInterpretation,"RGB") == 0
		&& vPlanarConfiguration == 0) {
		pnmstring="P6";
	}
	else {
		log << "Error - Only single channel or 3 channel RGB interleaved images supported" << endl;
		success=false;
	}

	if (success) {
		Assert(pnmstring);
		out << pnmstring << "\n";
		out << vColumns << " " << vRows << "\n";
		out << ((vBitsAllocated > 8 && vBitsStored < 8) ? 256 : (1<<vBitsStored)-1) << "\n";
	}


	Attribute *apixeldata=list[TagFromName(PixelData)];
	if (!apixeldata) {
		log << EMsgDC(MissingAttribute) << " - \"PixelData\"" << endl;
		success=false;
	}
	else if (!apixeldata->isOtherData()) {
		log << EMsgDC(PixelDataIncorrectVR) << endl;
		success=false;
	}

	if (success) {
		OtherUnspecifiedLargeAttributeBase *
			opixeldata = apixeldata->castToOtherData();
		Assert(opixeldata);


		// do NOT want to write all available bytes, since odd number will have ab even padding byte
		// and netpbm utilities freak if this is present with "Error reading magic number from Netpbm image stream.  Most often, this means your input file is empty." ...
		
		if (byteCount == 0) {
			// do it the "old" way, if for some reason we could not find a suitable value for something that contributes to the desired count; should never happen
			if (!quiet) log << "Writing VL " << dec << opixeldata->getVL() << " (dec) bytes ..." << endl;
			opixeldata->writeRaw(out);
		}
		else {
			if (!quiet) log << "Writing unpadded " << dec << byteCount << " (dec) bytes ..." << endl;
			opixeldata->writeRaw(out,0/*byte offset*/,byteCount);
		}
		
		if (!out.good()) {
			log << EMsgDC(Writefailed) << endl;
			success=false;
		}
	}

	return success ? 0 : 1;
}

	
